app/src/org/commcare/android/resource/AndroidResourceManager.java

```diff
@@ -0,0 +1,267 @@
+     * @param cancelCheckker allows resource installers to check if the
+     *                       launching process was cancelled
+     */
+    @Override
+    public void setUpgradeListeners(TableStateListener tableListener,
+                                    InstallCancelled cancelCheckker) {
+        super.setUpgradeListeners(tableListener, cancelCheckker);
+
+        tempUpgradeTable.setStateListener(tableListener);
+        tempUpgradeTable.setInstallCancellationChecker(cancelCheckker);
+    }
+
+    /**
+     * Save upgrade stats if the upgrade was cancelled and wasn't complete at
+     * that time.
+     */
+    public void upgradeCancelled() {
+        if (!isUpgradeTableStaged()) {
+            UpdateStatPersistence.saveStatsPersistently(app, updateStats);
+        } else {
+            Log.i(TAG, "Upgrade cancelled, but already finished with these stats");
+            Log.i(TAG, updateStats.toString());
+        }
+    }
+
+    public void incrementUpdateAttempts() {
+        updateStats.registerStagingAttempt();
+    }
+
+    /**
+     * Log update failure that occurs while trying to install the staged update table
+     */
+    public void recordUpdateInstallFailure(Exception exception) {
+        updateStats.registerUpdateException(exception);
+    }
+
+    public void recordUpdateInstallFailure(AppInstallStatus result) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+    }
+
+    /**
+     * Clear update table, log failure with update stats,
+     * and, if appropriate, schedule a update retry
+     *
+     * @param exception Log error with update statistics
+     * @param context   Used for showing pinned notification of update task retry
+     */
+    public void processUpdateFailure(Exception exception, Context context) {
+        updateStats.registerUpdateException(exception);
+
+        upgradeTable.clear();
+
+        retryUpdateOrGiveUp(context);
+    }
+
+    public void processUpdateFailure(AppInstallStatus result, Context ctx) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+
+        boolean reusePartialTable =
+                (result == AppInstallStatus.UnknownFailure ||
+                        result == AppInstallStatus.NoLocalStorage);
+
+        if (!reusePartialTable) {
+            upgradeTable.clear();
+        }
+
+        retryUpdateOrGiveUp(ctx);
+    }
+
+    private void retryUpdateOrGiveUp(Context ctx) {
+        if (updateStats.isUpgradeStale()) {
+            Log.i(TAG, "Stop trying to download update. Here are the update stats:");
+            Log.i(TAG, updateStats.toString());
+
+            UpdateStatPersistence.clearPersistedStats(app);
+
+            upgradeTable.clear();
+        } else {
+            Log.w(TAG, "Retrying auto-update");
+            UpdateStatPersistence.saveStatsPersistently(app, updateStats);
+            scheduleUpdateTask(ctx);
+        }
+    }
+
+    private void scheduleUpdateTask(final Context ctx) {
```

Comment:
might be worth naming or commenting this in a way that makes it clear that it's being used exclusively in the case of a retry after some sort of failure